md4c/md4c.c

```diff
@@ -3974,6 +4016,16 @@ md_process_inlines(MD_CTX* ctx, const MD_LINE* lines, int n_lines)
                         MD_LEAVE_SPAN(MD_SPAN_DEL, NULL);
                     break;
 
+                case '$':
+                    if(mark->flags & MD_MARK_OPENER) {
+                        MD_ENTER_SPAN((mark->end - off) % 2 ? MD_SPAN_LATEXMATH : MD_SPAN_LATEXMATH_DISPLAY, NULL);
+                        text_type = MD_TEXT_CODE; /* LaTeX should be read as code */
```

Comment:
I am wondering whether intorducing some `MD_TEXT_LATEXMATH` wouldn't be better. I can imagine some potential renderers may want to render differently from code.

Consider e.g. a HTML renderer which would pass the `MD_TEXT_LATEXMATH` text to some machinery which would use LaTeX to render the math into an image and embed that image in the HTML page.

Or a custom renderer which would (re)implement LaTeX math syntax and would support its rendering.

Reply:
The main reason I went with `MD_TEXT_CODE` is because there are some things that happen when setting code that apply to LaTeX code as well (like not escaping anything until the span closes). If the custom renderer wants to treat LaTeX code differently from any other code, it can still easily do that by setting some flag when entering an equation (similar to how it would have to watch for opening images to correctly process the alt attribute contents).

I can see there being benefits to not having to handle this in the renderer, but at the same time it would probably make the md4c code more convoluted (? I haven't looked at how the MD_TEXT_CODE is implemented). Let me know what you think, I'll update the PR to include the other changes :)

Reply:
The issue is that such renderer would have to test many conditions (is the span `MD_SPAN_LATEXMATH` **or** `MD_SPAN_LATEXMATH_DISPLAY`; and in the future also possibly some parent block if we introduce math blocks.

My 2nd argument is there is always just one parser (md4c.c), but there can potentially be many renderers, so it is better to place the complexity into the parser and make the renderers easier to write.

> I can see there being benefits to not having to handle this in the renderer, but at the same time it would probably make the md4c code more convoluted (? I haven't looked at how the MD_TEXT_CODE is implemented). Let me know what you think, I'll update the PR to include the other changes :)

I think that all would should be needed is to update the block at the end of `md_process_inlines()` which handles end of line. (Currently line number 4075 in upstream master.) 

Actually the question is whether this should be handled the same way or not.

And even if we stick with the current `MD_TEXT_CODE`, the assertions there are wrong right now: They verify we are indeed inside a code span denoted with `` ` `` chars so this:

```
this block $has a math span spread into
multiple lines$.
```

would likely explode on the assertion.

Reply:
> And even if we stick with the current `MD_TEXT_CODE`, the assertions there are wrong right > now: They verify we are indeed inside a code span denoted with `` ` `` chars [...]

That's a good point. I'll look into that, and introduce the additional type.

Reply:
Well, the 1st step here is to know how it should behave on EOLs.

Quite certainly it shouldn't be handled as a normal text where EOLs are translated into `MD_TEXT_BR` or `MD_TEXT_SOFTBR`. But IDK whether translating the EOL into a space as the code spans do is the right think for the math. What do you think?

Reply:
I think translating EOL to spaces is the right thing to do. The normal use-case for inline spans is to reference variables in text, so a very usual pattern is:
```
Bla bla $a$ bla bla...
```
and so $...EOL...$ is very unusual anyway.

Reply:
Ok. Then updating that block should be quite easy:

```
if(text_type == MD_TEXT_CODE  ||  text_type == _the_new_text_type_)
```

and replacing `MD_TEXT_CODE` with text_type in that branch. 

As of the assertions, I would simply remove them as they would become too complicated otherwise. 

Please also add some multi-line math into the test file so the branch is triggered by the tests.

Reply:
>  so $...EOL...$ is very unusual anyway.

We cannot rely on that. Even if long equations may be unlikely, many Markdown authors want to word wrap long lines at e.g. 80 characters. And the math span can begin at the offset 75 in the text.